File: prim_v2.cpp
#include<iostream>
#include<cstdio>
#include<algorithm>
#include<vector>
#include<stack>
#include<queue>
#include<climits>

#define MAX 5
#define Max(a,b) ((a>b)?a:b)
#define Min(a,b) ((a<b)?a:b)
#define FOR(i,n) for(int i=0;i<(n);i++)
#define S(n) (scanf("%d",&n))

using namespace std;

void _initialize()
{
}

int minKey(int key[],bool mstSet[])
{
    int mn = INT_MAX,minkey;
    for(int i=0;i<MAX;i++)
    {
        if(!mstSet[i] && key[i]<mn)
            mn = key[i],minkey = i;
    }
    return minkey;
}

void primMST(int graph[MAX][MAX])
{
    bool mstSet[MAX];
    int key[MAX],parent[MAX];
    FOR(i,MAX)
    {
        mstSet[i] = false;
        key[i] = INT_MAX;
    }
    key[0] = 0;
    parent[0] = -1;
    FOR(i,MAX-1)
    {
        int u = minKey(key,mstSet);
        mstSet[u] = true;
        FOR(v,MAX)
        {
            if(graph[u][v] && mstSet[v]==false && graph[u][v]<key[v])
            {
                key[v] = graph[u][v];
                parent[v] = u;
            }
        }
    }
    //for(int i=0;i<MAX;i++)
    //    cout<<mstSet[i]<<" ";
    cout<<"\n";
    for(int i=1;i<MAX;i++)
    {
        cout<<parent[i]<<" "<<i<<" "<<graph[i][parent[i]]<<"\n";
    }
}

int main()
{
    _initialize();
    int   graph[MAX][MAX]   =  {{ 0 ,   2 ,   0 ,   6 ,   0 } ,
                                { 2 ,   0 ,   3 ,   8 ,   5 } ,
                                { 0 ,   3 ,   0 ,   0 ,   7 } ,
                                { 6 ,   8 ,   0 ,   0 ,   9 } ,
                                { 0 ,   5 ,   7 ,   9 ,   0 }};
    primMST(graph);
}

File: primV2.py
#prim V2
from sys import maxint
MAX = maxint
def minKey(keys,mst):
	mn = MAX
	minkey = 0
	for i in range(len(keys)):
		if mst[i]==False and keys[i]<mn:
			mn = keys[i]
			minkey = i
	return minkey

def primMST(graph):
	l = len(graph)
	mst = []
	keys = []
	parent = []
	for i in range(l):
		keys.append(MAX)
		mst.append(False)
		parent.append(-1)
	keys[0] = 0
	parent[0] = -1
	for i in range(l-1):
		u = minKey(keys,mst)
		mst[u] = True
		for v in range(l):
			if graph[u][v]!=0 and mst[v]==False and graph[u][v]<keys[v]:
				keys[v] = graph[u][v]
				parent[v] = u
	for i in range(1,l):
		print "{0} - {1} : {2}".format(parent[i],i,graph[i][parent[i]])


if __name__ == "__main__":
	graph = [(0,2,0,6,0),(2,0,3,8,5),(0,3,0,0,7),(6,8,0,0,9),(0,5,7,9,0)]
	primMST(graph)
